odoo/addons/base/transpiler/js_transpiler.py

```diff
@@ -0,0 +1,467 @@
+"""
+This code is what let us uses ES6 like JS in odoo.
+Odoo expects all javascript to be divided into functions as "odoo.define(<module_name>, {...});".
+If any file has the /** odoo-module */ on top of it, it will get processed by this class.
+It performs several operations to get from ES6 syntax to the usual odoo one with minimal changes.
+This is done on the fly, this not a pre-processing tool.
+
+Caveat: This is done without a full parser, only using regex. One can only expect to cover as much edge cases
+as possible with reasonable limitations. Also, this only changes imports and exports, so all JS features used in
+the original source need to be supported by the browsers.
+"""
+
+import re
+from functools import partial
+
+def transpile_javascript(url, content):
+    """
+    Transpile the code from native JS modules to custom odoo modules.
+
+    :param content: The original source code
+    :param url: The url of the file in the project
+    :return: The transpiled source code
+    """
+    module_path = url_to_module_path(url)
+    legacy_odoo_define = get_aliased_odoo_define_content(module_path, content)
+
+    # The order of the operations does sometimes matter.
+    steps = [
+        convert_legacy_default_import,
+        convert_basic_import,
+        convert_default_import,
+        convert_star_import,
+        convert_unnamed_relative_import,
+        convert_from_export,
+        convert_star_from_export,
+        partial(convert_relative_require, url),
+        remove_index,
+        convert_export_function_or_class,
+        convert_variable_export,
+        convert_list_export,
+        convert_default,
+        partial(wrap_with_odoo_define, module_path),
+    ]
+    for s in steps:
+        content = s(content)
+    if legacy_odoo_define:
+        content += legacy_odoo_define
+    return content
+
+
+def url_to_module_path(url):
+    """
+    Odoo modules each have a name. (odoo.define("<the name>", function (require) {...});
+    It is used in to be required later. (const { something } = require("<the name>").
+    The transpiler transforms the url of the file in the project to this name.
+    It takes the module name and add a @ on the start of it, and map it to be the source of the static/src (or
+    static/tests) folder in that module.
+
+    in: web/static/src/one/two/three.js
+    out: @web/one/two/three.js
+    The module would therefore be defined and required by this path.
+
+    :param url: an url in the project
+    :return: a special path starting with @<module-name>.
+    """
+    match = re.match(r"""/?(?P<module>\w+)/([\w/]*/)?static/(?P<type>src|tests)(?P<url>/[\w/]*)""", url)
```

Comment:
Is this called often?
If so, compile it in a constant outside the function and use it:
```python
URL_RE = re.compile(r"""/?(?P<module>\w+)/([\w/]*/)?static/(?P<type>src|tests)(?P<url>/[\w/]*)""")

def url_to_module_path(url):
    match = URL_RE.match(url)
    if not match:
        raise ValueError(...)
    ...

Reply:
Note that `re` has its own cache, so under the assumption that these are called a lot, then not called for a while, then called a lot again, this would save memory: they'll be compiled and cached when they're hit, then as we load other PRs they'll get evicted from the cache and leave room for other stuff.

OTOH it means the COW can't reuse them, probably?